lib/customisation.nix

```diff
@@ -301,7 +301,26 @@ rec {
 
     in
     if missingArgs == { } then
-      makeOverridable f allArgs
+      if fargs ? mkPackage then
+        if fargs != { mkPackage = false; } then
+          # If your file is e.g. an attrset with a package and something else,
+          # use `{ callPackage }: { foo = callPackage ({ mkPackage}: ...); ... }`
+          # Explaining that in the error message would be too verbose and confusing.
+          # TODO: link to docs
+          throw ''
+            callPackage: A package function that uses `mkPackage` must only have `{ mkPackage }:` as its arguments.
+            Any other dependencies should be taken from the `mkPackage` callback.
+            Otherwise, such dependencies will not be overridable.
+
+            To illustrate, a dependency `foo` can be retrieved with:
+            { mkPackage }: mkPackage ({ layers, foo }: [
+              ...
+            ])
+          ''
+        else
+          f allArgs
```

Comment:
This avoids adding the second fixpoint (`callPackage`/`makeOverridable`/`.override`).

I should say that this is an overly broad interpretation of fixpoint, but a recursion is definitely there, so I'll just use the term. (even if `result` isn't passed back to `makeOverridable`'s `f`, the overriding dynamics are equally problematic)